rocAL/source/loaders/image/cifar10_loader_sharded.cpp

```diff
@@ -0,0 +1,166 @@
+THE SOFTWARE.
+*/
+
+#include "loaders/image/cifar10_loader_sharded.h"
+
+CIFAR10LoaderSharded::CIFAR10LoaderSharded(void* dev_resources) : _dev_resources(dev_resources) {
+    _loader_idx = 0;
+}
+
+void CIFAR10LoaderSharded::set_prefetch_queue_depth(size_t prefetch_queue_depth) {
+    if (prefetch_queue_depth <= 0)
+        THROW("Prefetch quque depth value cannot be zero or negative");
+    _prefetch_queue_depth = prefetch_queue_depth;
+}
+
+std::vector<std::string> CIFAR10LoaderSharded::get_id() {
+    if (!_initialized)
+        THROW("get_id() should be called after initialize() function");
+    return _loaders[_loader_idx]->get_id();
+}
+
+DecodedDataInfo CIFAR10LoaderSharded::get_decode_data_info() {
+    return _loaders[_loader_idx]->get_decode_data_info();
+}
+
+CIFAR10LoaderSharded::~CIFAR10LoaderSharded() {
+    _loaders.clear();
+}
+
+void CIFAR10LoaderSharded::fast_forward_through_empty_loaders() {
+    int loaders_count = _loaders.size();
+    // reject empty loaders and get to a loader that still has images to read
+    while (_loaders[_loader_idx]->remaining_count() == 0 && loaders_count-- > 0)
+        increment_loader_idx();
+}
+
+LoaderModuleStatus CIFAR10LoaderSharded::load_next() {
+    if (!_initialized)
+        return LoaderModuleStatus::NOT_INITIALIZED;
+
+    increment_loader_idx();
+
+    // Since loaders may have different number of images loaded, some run out earlier than other.
+    // Fast forward through loaders that are empty to get to a loader that is not empty.
+    fast_forward_through_empty_loaders();
+
+    auto ret = _loaders[_loader_idx]->load_next();
+
+    return ret;
+}
+
+void CIFAR10LoaderSharded::initialize(ReaderConfig reader_cfg, DecoderConfig decoder_cfg, RocalMemType mem_type,
+                                    unsigned batch_size, bool keep_orig_size) {
+    if (_initialized)
+        return;
+    _shard_count = reader_cfg.get_shard_count();
+    // Create loader modules
+    for (size_t i = 0; i < _shard_count; i++) {
+        std::shared_ptr loader = std::make_shared<CIFAR10Loader>(_dev_resources);
+        loader->set_prefetch_queue_depth(_prefetch_queue_depth);
+        _loaders.push_back(loader);
+    }
+    // Initialize loader modules
+    for (size_t idx = 0; idx < _shard_count; idx++) {
+        _loaders[idx]->set_output(_output_tensor);
+        _loaders[idx]->set_gpu_device_id(idx);
+        reader_cfg.set_shard_count(_shard_count);
+        reader_cfg.set_shard_id(idx);
+        _loaders[idx]->initialize(reader_cfg, decoder_cfg, mem_type, batch_size, keep_orig_size);
+    }
+    _initialized = true;
+}
+
+void CIFAR10LoaderSharded::start_loading() {
+    for (unsigned i = 0; i < _loaders.size(); i++) {
+        _loaders[i]->start_loading();
+        //  Changing thread scheduling policy and it's priority does not help on latest Ubuntu builds
+        //  and needs tweaking the Linux security settings , can be turned on for experimentation
+#if 0
```

Comment:
please combine both #if 0 blocks into 1 and add comments why it is commented out